src/osd/scrubber/pg_scrubber.cc

```diff
@@ -2776,24 +2777,38 @@ ReplicaReservations::no_reply_t::no_reply_t(
   ReplicaReservations& parent,
   std::string_view log_prfx)
     : m_osds{osds}
-    , m_conf{conf}
     , m_parent{parent}
     , m_log_prfx{log_prfx}
 {
-  using namespace std::chrono;
   auto now_is = clock::now();
-  auto timeout =
-    conf.get_val<std::chrono::milliseconds>("osd_scrub_reservation_timeout");
-
-  m_abort_callback = new LambdaContext([this, now_is]([[maybe_unused]] int r) {
-    // behave as if a REJECT was received
-    m_osds->clog->warn() << fmt::format(
-      "{} timeout on replica reservations (since {})", m_log_prfx, now_is);
-    m_parent.handle_no_reply_timeout();
-  });
+  auto timeout = conf.get_val<milliseconds>("osd_scrub_reservation_timeout");
+
+  m_abort_callback = new LambdaContext(
+      [this, now_is, pgid = m_parent.m_pg->pg_id]([[maybe_unused]] int r) {
+	// behave as if a REJECT was received
+	if (!m_abort_callback) {
+	  // already cancelled
+	  return;
+	}
+	m_abort_callback = nullptr;
+	m_osds->clog->warn() << fmt::format(
+	    "{} timeout on replica reservations (since {})", m_log_prfx,
+	    now_is);
+	PGRef pg = m_osds->osd->lookup_lock_pg(pgid);
```

Comment:
Normally, we just stash a reference to the pg in the callback and lock it.  Why is it preferable to look it up?

Reply:
I also don't see a check that the interval hasn't changed.

Reply:
Hmm - I'll need to ask you about that offline, to understand how is that handled if the PG was removed in the meantime (note that I only check for interval change in the receiving end of the message that
is triggered by the next line here)

Reply:
a) The callback would hold a reference, so it would still be live even if semantically deleted.  b) Prior to being removed from the OSD, there would have had to be an interval change making it no longer the primary.  From b), checking the interval should be sufficient even if the PG has been removed.  From a), the check is safe.